core/view-change.go

```diff
@@ -128,3 +153,71 @@ func validateMessageLog(replicaID uint32, view uint64, log messages.MessageLog,
 
 	return nil
 }
+
+func makeViewChangeProcessor(collect viewChangeCollector, processNewViewCert newViewCertProcessor) viewChangeProcessor {
+	var lock sync.Mutex
+
+	return func(vc messages.ViewChange) (new bool, _ error) {
+		lock.Lock()
+		defer lock.Unlock()
+
+		new, nvCert := collect(vc)
+		if nvCert != nil {
+			processNewViewCert(vc.NewView(), nvCert)
+		}
+
+		return
+	}
+}
+
+func makeViewChangeCollector(id, commitCertSize, n uint32) viewChangeCollector {
+	var (
+		view      uint64
+		collected = make(messages.NewViewCert, 1, n-commitCertSize+1)
+	)
+
+	return func(vc messages.ViewChange) (new bool, nvCert messages.NewViewCert) {
+		if newView := vc.NewView(); newView < view {
+			return false, nil
+		} else if newView > view {
+			collected = make(messages.NewViewCert, 1, cap(collected))
+			view = newView
+		}
+
+		if !isPrimary(view, id, n) {
+			return false, nil
+		}
+
+		if vc.ReplicaID() == id {
+			collected[0] = vc
+		} else if len(collected) <= int(n-commitCertSize) {
```

Comment:
This kind of check is called multiple times, so how about factoring out to a separate function with a descriptive name?

Reply:
Maybe just define `threshold := n-commitCertSize`?

Reply:
I think the condition here should be `len(collected) < int(n-commitCertSize)` resp. `len(collected) < int(threshold)`, need to check the unit-test...

Reply:
Ah, no, `commitCertSize == f+1`, so the condition is correct. Then we would define `threshold := n-commitCertSize+1`.

Reply:
Thank you for double checking, `threshold := n-commitCertSize` looks fine to me.

Reply:
Perhaps `viewChangeCertSize` would be a better name. What do you think about passing it to `makeViewChangeCertValidator`, `makeViewChangeCollector`, and `makeReqViewChangeCollector` as a parameter?

Reply:
Sounds nice 👍